Backend/FcTypeChecker.hs

```diff
@@ -222,6 +222,21 @@ tcType (FcTyApp ty1 ty2) = do
     _otherwise               -> throwError "tcType: Kind mismatch (FcTyApp)"
 tcType (FcTyCon tc) = lookupTyConKindM tc
 
+-- | Flatten out any or patterns in the alternatives
+flatAlts :: FcAlts 'Tc -> FcAlts 'Tc
+flatAlts = concatMap (\(FcAlt p rhs) -> [FcAlt p' rhs | p' <- flatPat p])
+
+-- | Flatten out any or patterns in the pattern
+flatPat :: FcPat 'Tc -> [FcPat 'Tc]
+flatPat (FcOrPat p1 p2)    = flatPat p1 ++ flatPat p2
+flatPat (FcVarPat x)       = [FcVarPat x]
+flatPat (FcConPatNs dc ps) = [FcConPatNs dc ps' | ps' <- cart (map flatPat ps)]
+
+-- | Gives every possible combination from the list of lists
+cart :: [[a]] -> [[a]]
+cart []       = pure []
+cart (xs:xss) = concat [[(x:ys) | ys <- cart xss] | x <- xs]
```

Comment:
Maybe move this one into `Utils.Utils` also? It seems too generic to be here.

Reply:
I agree